amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -313,8 +319,24 @@ private void errorOccurred(RecordFlow triggeringFlow, Throwable t) {
         }
     }
 
-    private void clearRecordsDeliveryQueue() {
+    // This method is not thread safe. This needs to be executed after acquiring lock on this.lockObject
+    private void clearRecordsDeliveryQueue(boolean isSubscriptionCompleting) {
```

Comment:
This method seems to be no-op if the delivery q is empty. Just check that in the beginning and return and continue if not.

    if(recordsDeliveryQueue.isEmpty()) {
      return;
    }
    // This will prevent further events from getting scheduled
    isAwaitingFinalAckForCurrentSubscription = true;
    recordsDeliveryQueue.clear();
    if(isSubscriptionCompleting) {
            try {
                // Wait for the configured time to get a notification for already delivered event, if any.
                lockObject.wait(TIME_TO_WAIT_FOR_FINAL_ACK_MILLIS);
            
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
Set isAwaitingFinalAckForCurrentSubscription to default value.
isAwaitingFinalAckForCurrentSubscription = false;

Reply:
Ack. Will make changes